src/resmom/mom_hook_func.c

```diff
@@ -3075,6 +3077,313 @@ record_job_last_hook_executed(unsigned int hook_event,
+				"execjob_end hook", wstat);
+			log_event(PBSEVENT_DEBUG2, PBS_EVENTCLASS_HOOK,
+				LOG_ERR, phook->hook_name, log_buffer);
+			hook_error_flag = 1;	/* hook results are invalid */
+	}
+
+	if (hook_error_flag == 0) {
+		/* hook results path */
+		snprintf(hook_outfile, MAXPATHLEN, FMT_HOOK_OUTFILE,
+			path_hooks_workdir,
+			HOOKSTR_EXECJOB_END,
+			phook->hook_name, mypid);
+
+		/* hook exited normally, get results from file  */
+		if (get_hook_results(hook_outfile, &accept_flag, &reject_flag,
+			reject_msg, sizeof(reject_msg), &reject_rerunjob,
+			&reject_deletejob, &reboot_flag, reboot_cmd,
+			HOOK_BUF_SIZE, &vnl_changes, pjob, phook, 0, NULL) != 0) {
+			log_event(PBSEVENT_DEBUG2, PBS_EVENTCLASS_HOOK,
+				LOG_ERR, phook->hook_name,
+				"Failed to get hook results");
+			vna_list_free(vnl_changes);
+			hook_error_flag = 1;
+		}
+	}
+
+	if ((hook_error_flag == 1) || (accept_flag == 0)) {
+
+		snprintf(log_buffer, sizeof(log_buffer),
+					"%s request rejected by '%s'",
+					hook_event_as_string(phook->event),
+					phook->hook_name);
+		log_event(PBSEVENT_DEBUG2, PBS_EVENTCLASS_HOOK,
+					LOG_ERR, phook->hook_name, log_buffer);
+
+		if ((reject_msg != NULL) && (reject_msg[0] != '\0')) {
+			snprintf(log_buffer, sizeof(log_buffer), "%s",
+				reject_msg);
+			/* log also the custom reject message */
+			log_event(log_type, log_class, LOG_ERR,
+				log_id, log_buffer);
+		}
+
+		/* Don't process anymore hooks on reject */
+		pjob = hook_input->pjob;
+
+		if (pjob->ji_qs.ji_svrflags & JOB_SVFLG_HERE) {
+#if MOM_ALPS
+			(void)mom_deljob_wait(pjob);
+
+			/*
+			* The delete job request from Server will have been
+			* or will be replied to and freed by the
+			* alps_cancel_reservation code in the sequence of
+			* functions started with the above call to
+			* mom_deljob_wait().  Set preq to NULL here so we
+			* don't try, mistakenly, to use it again.
+			*/
+			pjob->ji_preq = NULL;
+#else
+			numnodes = pjob->ji_numnodes;
+			preq = pjob->ji_preq;
+			pjob->ji_preq = NULL;
+			if (mom_deljob_wait(pjob) > 0) {
+				/* wait till sisters respond */
+				pjob->ji_preq = preq;
+			} else if (numnodes > 1) {
+				/*
+				* no messages sent, but there are sisters
+				* must be all down
+				*/
+				req_reject(PBSE_SISCOMM, 0, preq); /* all sis down */
+			} else {
+				reply_ack(preq);	/* no sisters, reply now  */
+			}
+#endif
+		} else {
+			void post_reply(job *, int);
+			post_reply(pjob, 0);
+			mom_deljob(pjob);
+		}
+		free(hook_input);
+		return;
+	}
+
+	if (reboot_flag) {
+		if (phook->user == HOOK_PBSUSER) {
+			log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_HOOK,
+				LOG_INFO, phook->hook_name,
+				"Not allowed to issue reboot if run as user");
+		} else {
+			log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_HOOK,
+				LOG_INFO, phook->hook_name,
+				"requested for host to be rebooted");
+			do_reboot(reboot_cmd);
+		}
+	}
+
+	/* Create task to check and run next hook script */
+	new_task = set_task(WORK_Immed, 0,
```

Comment:
I think we are leaking new_task here. I don't see run_execjob_end_hooks freeing it. Also, if you choose to free it in the callback function then you will have to change mom_process_hook as well because there you are passing stack variable.

Reply:
No, we are not leaking new_task here, this will free in dispatch_task function.